naturalcrit/homebrewery
# Intersection Observers for getting "Current Page(s)"

This PR is mostly motivated by the work on the #3799 PR which allows users to view many pages at once, but was causing issues with what is the "current page" and also issues of viewing "page 10 out of 9 total pages"-like issues in the Toolbar.

## Description

This sets up two Intersection Observers: the first captures every page that is at least 30% visible inside the `.pages` container, and the second captures every page that has at least one pixel on the horizontal center line of `.pages`.  Both can be arrays of integers (page index).

The `visiblePages` array is duplicated and formatted into a `formattedPages` state variable as strings, which gets displayed in the toolbar. `visiblePages` is kept so that it can be used in the "previous/next page" buttons of the toolbar later.

The toolbar displays the `formattedPages` strings, unless the user clicks into the page input and enters their own integer (only a single integer, no range), which can then jump the preview to that page on Enter or blur().

The Arrow 'change page' buttons jump the preview back and forth by a 'full set'.  Meaning that if you are viewing one page a time it will move through the previous/next pages one at a time, but if you are viewing 10 pages at once it will move you through 10 pages at a time.

## Related Issues or Discussions

- Closes #3824
- #3694 -- ultimately scuttled for another PR but raises issue of bad page count calculations

## QA Instructions, Screenshots, Recordings

First, sorry for the bad single commit on this one, it doesn't really walk you through the development.

An example document (500 pages) with some additional styling to make it easy to see page numbers:

[HB-NecromanticArmorofSaladDressing.txt](https://github.com/user-attachments/files/17656392/HB-NecromanticArmorofSaladDressing.1.txt)

**or**, just put this CSS in any long brew:
```css
.page:has(.insideCover, .frontCover, .backCover)::after {
  display: grid;
}

.page::after {
  all:unset;
  content: counter(page-numbers);
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: #AAAA;
  display: grid;
  place-items: center;
  font-size: 200px;
  }
  
.page:nth-child(even):after {
  transform: unset;
 }

```

To Test:

1. Load the page and it should load up with `1 / [total pages]` in the toolbar.
2. You should see it change when scrolling in the Preview pane, stopping at `1` or the total page count without going over.
3. Then, change the zoom so you see multiple pages and you should see *ranges* of pages in that input, like `10 - 12`.  You can also change the current "Spread" to facing or flow.  
4. Try the "previous/next" arrows. It should not overrun past the limits.
5. Try typing a single integer in the 'current page' input and hitting `enter` or clicking outside of it-- it should take you to that page.  Typing non-integer characters should be ignored, as well as ranges, and integers outside the total page count will bring you to the last page.
7. After entering your own page number, scrolling again continues to work.

Single page with >30% of area inside view:
<img width="1240" alt="image" src="https://github.com/user-attachments/assets/efa22199-bb97-48f6-b9c1-30912dd37d52">

Book spread, shows that only pages with >30% of area inside view are shown in the 'current pages' input:
<img width="1058" alt="image" src="https://github.com/user-attachments/assets/291b2f8b-c52c-4ebb-b5b2-425530d9ac32">

Many pages, shows the "range":
<img width="1058" alt="image" src="https://github.com/user-attachments/assets/4f422b8a-33f0-40d9-9942-ede699d9a38d">


### Reviewer Checklist

**No checklist yet. (I have to go to bed)**

*Reviewers, refer to this list when testing features, or suggest new items *
  - [ ] Page ranges are shown when multiple pages in view (more than a sliver).
  - [ ] Prev/Next Page buttons in toolbar move through based on how many pages are in view.
  - [ ] Accurately describes pages in view when pages are in `flex` format (multiple pages in row)
  - [ ] Brew/Source Jump buttons still work
  - [ ] Prev/Next buttons in toolbar still work
  - [ ] Does not show current page higher than total number of pages, or lower than "1". 
  - [ ] Functions at any zoom level
  - [ ] Doesn't cause performance issues

<details><summary>Copy this list</summary>

```
  - [ ] Page ranges are shown when multiple pages in view (more than a sliver).
  - [ ] Prev/Next Page buttons in toolbar move through based on how many pages are in view.
  - [ ] Accurately describes pages in view when pages are in `flex` format (multiple pages in row)
  - [ ] Brew/Source Jump buttons still work
  - [ ] Prev/Next buttons in toolbar still work
  - [ ] Does not show current page higher than total number of pages, or lower than "1". 
  - [ ] Functions at any zoom level
  - [ ] Doesn't cause performance issues
```

</details>
